adding hill-climbing smooth2 option for in msh clean (ds = 2)

WPringle opened the pull request:
- adding hill-climbing smoother, smooth2 (from mesh2d) as option 2 for the 'ds' cleaner. This can be used to avoid the situations where the implicit smoother can completely shift nodes that we may not want to be shifted. 

In this example we pave over the whole islands and use the edgefx that is calculated based on the local mean shoreline ('lmsl' edgefx option). No fixed points are used so direct smoother can freely move any vertex within the domain, although we do not wish for them to be shifted from our lmsl. It would appear that smooth2 avoids this kind of shifting:

m = m.clean('ds',0); % no smoother
![image](https://user-images.githubusercontent.com/21131934/82975967-85b5ef00-9fa3-11ea-8a0e-5982f25eed4b.png)

m = m.clean('ds',1); % using implicit smoother
![image](https://user-images.githubusercontent.com/21131934/82975914-63bc6c80-9fa3-11ea-9f2e-6294c53d4d3b.png)

m = m.clean('ds',2); % smooth2 hill-climbing
![image](https://user-images.githubusercontent.com/21131934/82976123-e5ac9580-9fa3-11ea-993b-b07931ecc19a.png)

WPringle committed:
adding smooth2 option for ds

- adding smooth2 from mesh2d as option 2 for the 'ds' cleaner. This can be used to avoid the situations where the implicit smoother can completely shift nodes that should not be shifted etc.

krober10nd commented:
I modified smooth2 years ago. It’s already in utilities. This figure you
show doesn’t make sense

On Wed, 27 May 2020 at 12:55 AM William Pringle <notifications@github.com>
wrote:

>
>    - adding smooth2 from mesh2d as option 2 for the 'ds' cleaner. This
>    can be used to avoid the situations where the implicit smoother can
>    completely shift nodes that should not be shifted etc.
>    - requires user to download Mesh2D separately (similar to m_map).
>
> Example:
> m = m.clean('ds',0); % no smoother
> [image: image]
> <https://user-images.githubusercontent.com/21131934/82975967-85b5ef00-9fa3-11ea-8a0e-5982f25eed4b.png>
>
> m = m.clean('ds',1); % using implicit smoother
> [image: image]
> <https://user-images.githubusercontent.com/21131934/82975914-63bc6c80-9fa3-11ea-9f2e-6294c53d4d3b.png>
>
> m = m.clean('ds',2); % smooth2 mesh2d smoother
> [image: image]
> <https://user-images.githubusercontent.com/21131934/82976123-e5ac9580-9fa3-11ea-993b-b07931ecc19a.png>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/CHLNDDEV/OceanMesh2D/pull/87
> Commit Summary
>
>    - adding smooth2 option for ds
>
> File Changes
>
>    - *M* @msh/msh.m
>    <https://github.com/CHLNDDEV/OceanMesh2D/pull/87/files#diff-2eaf8271fb937bb71262fad476427155>
>    (11)
>
> Patch Links:
>
>    - https://github.com/CHLNDDEV/OceanMesh2D/pull/87.patch
>    - https://github.com/CHLNDDEV/OceanMesh2D/pull/87.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CHLNDDEV/OceanMesh2D/pull/87>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOBZ7HL6DTWT3FXPOBH573RTSFMFANCNFSM4NLWS3HQ>
> .
>
-- 
--------------------
Keith Roberts
Post-doctoral researcher
University of São Paulo
krober@usp.br <krober10@nd.edu>

WPringle commented:
Keith,
This is not a modification of smooth2 and smooth2 is not in the utilities if you care to check. This is an option to use smooth2 instead of the direct_smoother_lur.m function.

What doesn't make sense in the above figure? It clearly shows using ds =2 (smooth2) does not shift nodes from the original shoreline position but using ds = 1 does (direct_smoother_lur).

krober10nd commented:
"requires user to download Mesh2D separately (similar to m_map)."

Why do you need to download mesh2d to use a laplacian smoother? 

"This is not a modification of smooth2 and smooth2 is not in the utilities if you care to check. This is an option to use smooth2 instead of the direct_smoother_lur.m function."

https://github.com/CHLNDDEV/OceanMesh2D/blob/Projection/utilities/smoothmesh.m

"What doesn't make sense in the above figure? It clearly shows using ds =2 (smooth2) does not shift nodes from the original shoreline position but using ds = 1 does (direct_smoother_lur)."

The direct smoother has been used for years with great success. Why suddenly does it stop working?

WPringle commented:
Smooth2 is not a laplacian smoother and smoothmesh is not a modification of smooth2. Smooth2 is the hill-climbing algorithm actually more like our mesh generation. It resulted in a non-trivial overall improvement to the mesh quality.  

Yes I know. Normally we just mesh the oceanside so that mesh boundary which is the shoreline is never moved. The problem appears because in this example we pave over the whole island and use the 'lmsl' option. Unfortunately in this case the direc_smoother_lur did the shifting. This is not something I thought would happen either but it did. So why not have another smoother option for cleaning if required?

WPringle committed:
adding smooth2 to utilities and calling inpoly in deltri2

krober10nd commented:
I see. It's interesting. Can you provide the mesh depicted in the image before the application of the direct smoother so we can see why this may happen?

WPringle committed:
correcting comments and help

krober10nd commented:
It appears to be shifted by a constant as well.

WPringle commented:
OK so I found out that we already had all the required files in the GEOM_UTIL so I added the smooth2 function to utilities and edited a call to inpoly2 to inpoly. This avoids requirement to download mesh2d. I also changed to highlight that it is a hill-climbing smoother. 

I've attached the script and the mesh here
[Guam_Mesh.zip](https://github.com/CHLNDDEV/OceanMesh2D/files/4689664/Guam_Mesh.zip)
I should mention that to run the first part of the script you need the geodata_improv branch to handle the extractContour call correctly.

krober10nd commented:
I couldn't run the example because I didn't have the data files. However, I did run just direct smoother (on its own) on the mesh and I did not notice any shift in the islands. By default it locks boundaries (treats them as fixed points inside the direct smoother). Perhaps the fixed points depicted in blue did not get converted to the correct projected space and were not constrained?

WPringle commented:
> I couldn't run the example because I didn't have the data files. However, I did run just direct smoother (on its own) on the mesh and I did not notice any shift in the islands. By default it locks boundaries (treats them as fixed points inside the direct smoother). Perhaps the fixed points depicted in blue did not get converted to the correct projected space and were not constrained?

The files are easily downloadable from noaa database. But yes the mesh I provided was already smoothed and had overland removed so it doesn't do anything bad. The problem was the mesh that comes out of meshgen. I didn't save that mesh. I'll run the script through again and give you that mesh.  

Note the blue points are not fixed points! No fixed points here. Blue is simply plotting the lmsl shoreline.

WPringle commented:
OK I did a reduced example to quickly get a result using min_el of 1 km. It's not as obvious but still there is a shift.
no clean, ds = 1, ds = 2
![image](https://user-images.githubusercontent.com/21131934/83045616-6a7eca00-a00b-11ea-97bf-b2dbfabf2da0.png)

Try it out on this same uncleaned mesh. Edited to include the 'lmsl' geodata. 
[Guam_Mesh_uncleaned.zip](https://github.com/CHLNDDEV/OceanMesh2D/files/4690099/Guam_Mesh_uncleaned.zip)

For plotting I do:
plot(m)
plot(lmsl,[],[],1); % plots shoreline on top of the mesh figure

krober10nd commented:
I see. yea of course there's a shift, they're not fixed.

 No optimizer will constrain points if they're not specified as immovable. The fact that smooth2 moves less doesn't mean it's a fix, however it just moves them less.

WPringle commented:
> I see. yea of course there's a shift, they're not fixed.
> 
> No optimizer will constrain points if they're not specified as immovable. The fact that smooth2 moves less doesn't mean it's a fix, however it just moves them less.

I'm not saying direct smoother is bad it just doesn't work as we would like in this situation.

For this example smooth2 works way better and there is no shift. User can decide which to use. I would use smooth2 of course.

krober10nd commented:
Of course. Ok, that's fine. The direct smoother improves the quality of the mesh respecting user defined fixed points.

krober10nd commented:
The OP was a bit confusing as you stated the nodes should not be shifted in blue. If nobody is constraining them, then why should they not be shifted?

WPringle commented:
> The OP was a bit confusing as you stated the nodes should not be shifted in blue. If nobody is constraining them, then why should they not be shifted?

Agreed, I have tried to make the OP clearer.

krober10nd commented:
Cool, thanks. I think smooth2 preserves the mean mesh size if no mesh size function is given. Because of this the high resolution areas around the shoreline are preserved.

krober10nd reviewed (approved):
lgtm
